File: MPlayerX/PlayerCore.m
/*
 * MPlayerX - PlayerCore.m
 *
 * Copyright (C) 2009 - 2011, Zongyao QU
 * 
 * MPlayerX is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * MPlayerX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with MPlayerX; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
#import "PlayerCore.h"
#import "CocoaAppendix.h"

#define kPlayerCoreTermNormal		(0)

@interface PlayerCore (PlayerCoreInternal)
-(void) readOutput:(NSNotification*)notification;
-(void) readError:(NSNotification*)notification;
-(void) taskHasTerminated:(NSNotification*)notification;
@end

@implementation PlayerCore

@synthesize delegate;

#pragma mark Init/Dealloc
-(id) init 
{
	self = [super init];
	
	if (self) {
		delegate = nil;
		task = nil;
		runningModes = [[NSArray alloc] initWithObjects:NSDefaultRunLoopMode, NSModalPanelRunLoopMode, NSEventTrackingRunLoopMode, nil];
	}
	return self;
}

-(void) dealloc
{
	// The terminate function calls the delegate's functions, which could cause logic errors
	delegate = nil;

	[self terminate];
	[runningModes release];
	
	[super dealloc];
}

#pragma mark Function
- (void) terminate
{
	if (task) {
		// To prevent the function from running multiple times
		NSTask *backup = task;
		task = nil;
		
		if ([backup isRunning]) {
			[backup terminate];
			[backup waitUntilExit];
		}
		[backup release];
	}
}

- (BOOL) playMedia: (NSString *) moviePath withExec: (NSString *) execPath withParams: (NSArray *) params
{
	if (moviePath && execPath) {
	
		[self terminate];
		
		// Create the task
		task = [[NSTask alloc] init];
		// Associate input/output
		[task setStandardInput:[NSPipe pipe]];
		[task setStandardOutput:[NSPipe pipe]];
		[task setStandardError: [NSPipe pipe]];
		// Specify the path of the exec to run
		[task setLaunchPath: execPath];
		// Create argv
		if (params) {
			[task setArguments: [params arrayByAddingObject:moviePath]];
		} else {
			[task setArguments: [NSArray arrayWithObject:moviePath]];
		}
		
		MPLog(@"%@", [[task arguments] componentsJoinedByString:@"\n"]);
		
		// Set environment parameters
		NSMutableDictionary *env = [[[NSProcessInfo processInfo] environment] mutableCopy];
		[env setObject:@"1" forKey:@"DYLD_BIND_AT_LAUNCH"]; //delete the message for DYLD
		[env setObject:@"xterm" forKey:@"TERM"]; // delete the message from mplayer about the "unknown" terminal
		[env removeObjectForKey:@"MPLAYER_HOME"];
		[env removeObjectForKey:@"HOME"];
		[task setEnvironment:env];
		[env autorelease];
		
		[task setCurrentDirectoryPath:[execPath stringByDeletingLastPathComponent]];
		
		// Set up the notification/listening mechanism
		[[NSNotificationCenter defaultCenter] addObserver:self
												 selector:@selector(readOutput:)
													 name:NSFileHandleReadCompletionNotification
												   object:[[task standardOutput] fileHandleForReading]];
		[[NSNotificationCenter defaultCenter] addObserver:self
												 selector:@selector(readError:)
													 name:NSFileHandleReadCompletionNotification
												   object:[[task standardError] fileHandleForReading]];
		
		[[NSNotificationCenter defaultCenter] addObserver:self
												 selector:@selector(taskHasTerminated:)
													 name:NSTaskDidTerminateNotification
												   object:task];
		
		[[[task standardOutput] fileHandleForReading] readInBackgroundAndNotifyForModes:runningModes];
		[[[task  standardError] fileHandleForReading] readInBackgroundAndNotifyForModes:runningModes];

		// Launch the task
		[task launch];
		return YES;
	}
	return NO;
}

- (BOOL) sendStringCommand: (NSString *) cmd
{
	if (task && [task isRunning]) {
		// MPLog(@"%@",cmd);
		[[[task standardInput] fileHandleForWriting] writeData:[cmd dataUsingEncoding:NSUTF8StringEncoding]];
		return YES;
	}
	return NO;
}

#pragma mark Internal 
-(void) readOutput:(NSNotification*)notification
{
	if (task && [task isRunning]) {
		NSData *data = [[notification userInfo] objectForKey:NSFileHandleNotificationDataItem];
		
		if (([data length] != 0) && delegate) {
			[delegate playerCore:self outputAvailable:data];
		}
		[[[task standardOutput] fileHandleForReading] readInBackgroundAndNotifyForModes:runningModes];
	}
}

-(void) readError:(NSNotification*)notification
{
	if (task && [task isRunning]) {
		NSData *data = [[notification userInfo] objectForKey:NSFileHandleNotificationDataItem];

		if (([data length] != 0) && delegate) {
			[delegate playerCore:self errorHappened:data];
		}
		[[[task  standardError] fileHandleForReading] readInBackgroundAndNotifyForModes:runningModes];
	}
}

-(void) taskHasTerminated:(NSNotification*)notification
{
	[[NSNotificationCenter defaultCenter] removeObserver:self];

	// Get the return status; 0 means normal exit
	// At this point the task variable may have become nil
	if (delegate) {
		[delegate playerCore:self hasTerminated:([[notification object] terminationStatus] != kPlayerCoreTermNormal)];
	}
}
@end

File: MPlayerX/PlayerCore.swift
/*
 * MPlayerX - PlayerCore.swift
 */

import Cocoa

private let kPlayerCoreTermNormal: Int32 = 0

@objc protocol PlayerCoreDelegate: NSObjectProtocol {
    func playerCore(_ player: Any, hasTerminated byForce: Bool)		/**< notifies that the playback task has terminated */
    func playerCore(_ player: Any, outputAvailable outData: Data)		/**< output is available */
    func playerCore(_ player: Any, errorHappened errData: Data)		/**< error output is available */
}

@objc(PlayerCore)
class PlayerCore: NSObject {
    @objc weak var delegate: PlayerCoreDelegate?

    private var task: Process?
    private let runningModes: [RunLoop.Mode] = [.default, .modalPanel, .eventTracking]

    deinit {
        delegate = nil
        terminate()
    }

    @objc func terminate() {
        guard let currentTask = task else { return }
        task = nil

        if currentTask.isRunning {
            currentTask.terminate()
            currentTask.waitUntilExit()
        }
    }

    @objc(playMedia:withExec:withParams:)
    func playMedia(_ moviePath: String?, withExec execPath: String?, withParams params: [Any]?) -> Bool {
        guard let moviePath = moviePath, let execPath = execPath else { return false }

        terminate()

        let newTask = Process()
        newTask.standardInput = Pipe()
        newTask.standardOutput = Pipe()
        newTask.standardError = Pipe()
        newTask.launchPath = execPath

        let stringParams = params?.compactMap { $0 as? String } ?? []
        newTask.arguments = stringParams + [moviePath]

        MPLogString((newTask.arguments ?? []).joined(separator: "\n"))

        var env = ProcessInfo.processInfo.environment
        env["DYLD_BIND_AT_LAUNCH"] = "1"
        env["TERM"] = "xterm"
        env.removeValue(forKey: "MPLAYER_HOME")
        env.removeValue(forKey: "HOME")
        newTask.environment = env
        newTask.currentDirectoryPath = (execPath as NSString).deletingLastPathComponent

        task = newTask

        let outputHandle = (newTask.standardOutput as? Pipe)?.fileHandleForReading
        let errorHandle = (newTask.standardError as? Pipe)?.fileHandleForReading

        if let outputHandle = outputHandle {
            NotificationCenter.default.addObserver(self, selector: #selector(readOutput(_:)), name: FileHandle.readCompletionNotification, object: outputHandle)
        }
        if let errorHandle = errorHandle {
            NotificationCenter.default.addObserver(self, selector: #selector(readError(_:)), name: FileHandle.readCompletionNotification, object: errorHandle)
        }
        NotificationCenter.default.addObserver(self, selector: #selector(taskHasTerminated(_:)), name: Process.didTerminateNotification, object: newTask)

        outputHandle?.readInBackgroundAndNotify(forModes: runningModes)
        errorHandle?.readInBackgroundAndNotify(forModes: runningModes)

        newTask.launch()
        return true
    }

    @objc(sendStringCommand:)
    func sendStringCommand(_ cmd: String?) -> Bool {
        guard let cmd = cmd,
              let currentTask = task,
              currentTask.isRunning,
              let inputPipe = currentTask.standardInput as? Pipe,
              let data = cmd.data(using: .utf8) else {
            return false
        }

        inputPipe.fileHandleForWriting.write(data)
        return true
    }

    @objc private func readOutput(_ notification: Notification) {
        guard let currentTask = task, currentTask.isRunning else { return }
        let data = notification.userInfo?[NSFileHandleNotificationDataItem] as? Data ?? Data()

        if !data.isEmpty {
            delegate?.playerCore(self, outputAvailable: data)
        }

        (currentTask.standardOutput as? Pipe)?.fileHandleForReading.readInBackgroundAndNotify(forModes: runningModes)
    }

    @objc private func readError(_ notification: Notification) {
        guard let currentTask = task, currentTask.isRunning else { return }
        let data = notification.userInfo?[NSFileHandleNotificationDataItem] as? Data ?? Data()

        if !data.isEmpty {
            delegate?.playerCore(self, errorHappened: data)
        }

        (currentTask.standardError as? Pipe)?.fileHandleForReading.readInBackgroundAndNotify(forModes: runningModes)
    }

    @objc private func taskHasTerminated(_ notification: Notification) {
        NotificationCenter.default.removeObserver(self)
        let process = notification.object as? Process
        delegate?.playerCore(self, hasTerminated: (process?.terminationStatus ?? kPlayerCoreTermNormal) != kPlayerCoreTermNormal)
    }
}
